### ISSUE TITLE
Rate limit totality of INat API calls 

### ISSUE THREAD

**Issue by JoeCohen** [MushroomObserver/mushroom-observer#2320]
Rate limit iNat API calls

- [ ] Rate limit for the totality of requests (all users) handled by all background jobs.

**Comment by JoeCohen** [comment]
On Wed, Jan 15, 2025 at 8:10 AM Nathan Wilson <[nathan@collectivesource.com](mailto:nathan@collectivesource.com)> wrote:
TL;DR it's more complicated than that.

I'd store it in the database as a model so the accounting doesn't get screwed up by deployments.  I think you'll also need to do that to get it to work with the job system since those are separate processes and hence would have their own instances of a PORO.  You'll need to be careful to avoid race conditions.  Each job could request a conceptual iNat token and get back an expected run time.  It also depends on how strictly iNat enforces the 60 per minute limit.  For example, if we don't have any traffic for an hour, but then we want 61 calls immediately, can we just do them or do we have to do 60 in the next minute and 1 in the minute after that?  It's likely best to assign it a runtime that is 1 second in the future for every unresolved token.  However, this could be an issue if 1000 are queued and then our system is down for an hour.  Not clear what happens to jobs in that circumstance.  SolidQueue doesn't seem to pick up on stuff that was in the queue before a server restart.  That's something that would be worth sorting out.

> On Wed, Jan 15, 2025 at 10:38 AM Joseph Cohen <[jdcohenesq@gmail.com](mailto:jdcohenesq@gmail.com)> wrote:
Question: How can I rate-limit outgoing calls to iNat's API. (via email so that jason can react if he wants)
>
> Inat wants us to limit calls to its API to 60/minute.
This is a roadblock to letting MO users import all their iNat obss at once.
(There could be hundreds of iNat obss per MO users, and multiple MO users could be trying to import at once.)
(And -- looking ahead --it gets worse when we implement exporting MO Obss to iNat.)
>
> My idea:
Channel all iNat API requests to a single (PORO) class method. 
The class includes a class variable which stores the time when the iNatAPI was last called. Something like @@last_called.
If that time is < 1.second.ago, then sleep(time.now -  @@last_called).
> \- Joe
